### ISSUE TITLE
Private Network Access

### ISSUE THREAD

**Issue by lidel** [ipfs/in-web-browsers#184]
This is a placeholder issue for tracking the work which right now lives as a W3C draft at https://wicg.github.io/private-network-access/

##  Summary

The proposal is to deny access to  LAN and loopback addresses by default, and allow them only when additional CORS flag `Access-Control-Request-Private-Network: true` is sent by the server's response during preflight response. 

This is a good thing, because it closes security issue of websites probing insecure services with default admin passwords etc.

## Impact on localhost IPFS gateway

If this gets implemented by browser vendors (very likely, as its championed by Google eng.), websites won't be able to access local HTTP ports: API and Gateway unless the gateway responds with `Access-Control-Request-Private-Network: true` to the preflight request. 

This can be easily added to the gateway logic, however there are some unknowns that depend on vendor implementation, and require additional QA, namely:
- on-the-fly redirects applied by ipfs-companion browser extension via 
- `ipfs://` requests made after registering local gateway as handler via  [registerProtocolHandler](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler) API

**Comment by lidel** [comment]
Intent to Prototype: Private Network Access permission to relax mixed content:
- https://groups.google.com/a/chromium.org/g/blink-dev/c/6MczoSFGiHo/m/IigYuhu7AwAJ

IPFS would benefit from a way to lift mixed-content for trusted  LANs. For example, it would allow JS nodes (and [@helia/verified-fetch](https://www.npmjs.com/package/@helia/verified-fetch)) to opportunistically use a LAN IPFS nodes as a local block cache, when present.

**Comment by lidel** [comment]
Firefox Intent to Prototype: Local Network Access (LNA) Defense via Permission Prompts
- https://groups.google.com/a/mozilla.org/g/dev-platform/c/B8oN3ARp_j0/m/rWKXmnj4AAAJ?pli=1
  - https://bugzilla.mozilla.org/show_bug.cgi?id=1481298
  - https://docs.google.com/document/d/1n0kKxt9pS9qDlu_9i5W8IXA594r4pUOKmN9H35cZ8j0/edit?tab=t.0

**Comment by lidel** [comment]
Webkit
- https://bugs.webkit.org/show_bug.cgi?id=250607

**Comment by lidel** [comment]
- cc potential use case here is allowing users to grant  web apps, browser extensions or service worker gateway access to a local node user is already running https://github.com/ipfs/service-worker-gateway/issues/299 (currently helia filters out "unsafe" and local multiaddrs to avoid mixed-context errors)
- https://github.com/WICG/local-network-access → https://wicg.github.io/local-network-access

**Comment by lidel** [comment]
https://developer.chrome.com/blog/local-network-access:
> ### Our initial version of **Local Network Access is ready for opt-in testing in Chrome 138**.
>  Users can enable the new permission prompt by setting chrome://flags#local-network-access-check to "Enabled (Blocking)". This supports triggering the Local Network Access permission prompt for requests initiated using the JavaScript fetch() API, subresource loading, and subframe navigation.
>
> A demo site is available at https://lna-testing.notyetsecure.com/ for triggering different forms of local network requests.

> ### Adapting your website for new Local Network Access restrictions in Chrome:
> https://docs.google.com/document/d/1QQkqehw8umtAgz5z0um7THx-aoU251p705FbIQjDuGs/edit?tab=t.0#heading=h.v8oobsqxbxxy

**Comment by lidel** [comment]
Did some research this weekend, mainly to understand if there are breaking changes coming and if we have to do anything.

## TL;DR

Things still work even if we don't implement any new headers. The third-party app `https://webui.ipfs.io` can talk to local Kubo RPC in Chrome 145. Users see a one-time permission prompt:

> <img width="433" height="220" alt="Image" src="https://github.com/user-attachments/assets/95adb837-8013-4505-b61f-8bc83a113e5f" />

The prompt shows a generic device name. With `Private-Network-Access-Name` headers, it could show `ipfs-kubo-rpc-k51qzi5uqu5d...` instead - but Chrome doesn't display this header yet.

## Proof-of-concept for Kubo

I prototyped PNA headers for both HTTP endpoints in Kubo:

**RPC API headers:**

```
Access-Control-Allow-Private-Network: true
Private-Network-Access-Name: ipfs-kubo-rpc-k51qzi5uqu5dgjc...
Private-Network-Access-ID: 01:23:45:67:89:ab 
```

**Gateway headers:**

```
Access-Control-Allow-Private-Network: true
Private-Network-Access-Name: ipfs-gateway-k51qzi5uqu5dgjc...
Private-Network-Access-ID: 01:23:45:67:89:ab
```

The Name uses the node's PeerID (base36) for identification.
The ID is MAC-like hex we derive from peerid. It has a nice property of persisting permission to access specific HTTP API backend by specific PeerID even if IP changes (e.g. in LAN, due to DHCP). Nice quality of life, but also not a deal breaker, people rarely change IPs/put persistent services on DHCP.

**No benefit in merging yet:**

1. Chrome doesn't display `Private-Network-Access-Name` in the permission UI
2. Upstream libraries (`rs/cors`, `go-ipfs-cmds`, `boxo/gateway`) lack PNA support
3. Browser-side `targetAddressSpace` fetch option isn't standardized

## Client-side: `targetAddressSpace`

Browsers will at some point  add a `fetch()` option to declare local network intent:

```javascript
// Kubo RPC
fetch("http://127.0.0.1:5001/api/v0/id", {
  targetAddressSpace: "loopback"
})

// LAN gateway (for Helia browser nodes)
fetch("http://192.168.1.100:8080/ipfs/bafy...", {
  targetAddressSpace: "local"
})
```

I suspect this will trigger correct use of `Private-Network-Access-*` headers in browser UI  AND also enable exception for mixed-content to LAN boxes.

IIUC this (PNA preflights) is not really ready yet , Chrome did not show name in UI.

## Next steps

1. Monitor Chrome's `Private-Network-Access-Name` support
2. Ship Kubo with extra headers once Chrome starts displaying device names in UI (low priority) or if Helia roadmap includes access to local nodes (tbd, for now also priority, Helia ignores local peers by default atm)
3. Update `fetch` in  `js-kubo-rpc-client` / verified fetch / helia http retrieval clients  to explicitly use PNA, when `targetAddressSpace` is standardized

## References

- [Chrome Local Network Access](https://developer.chrome.com/blog/local-network-access)
- [WICG spec (draft)](https://wicg.github.io/local-network-access/)
- [PNA Permission Prompt](https://developer.chrome.com/blog/pna-permission-prompt-ot-end)
